src/hotspot/share/runtime/flags/jvmFlagAccess.cpp

```diff
@@ -0,0 +1,397 @@
+                          "[ %f ... %f ]\n",
+                        name, value, min, max);
+  }
+  JVMFlag::Error typed_check_constraint(void* func, double value, bool verbose) const {
+    return ((JVMFlagConstraintFunc_double)func)(value, verbose);
+  }
+  void print_range_impl(outputStream* st, double min, double max) const {
+    st->print("[ %-25.3f ... %25.3f ]", min, max);
+  }
+  void print_default_range(outputStream* st) const {
+    st->print("[ %-25.3f ... %25.3f ]", DBL_MIN, DBL_MAX);
+  }
+};
+
+#define FLAG_ACCESS_IMPL_INIT(t) \
+  static FlagAccessImpl_ ## t   flag_access_ ## t;
+
+#define FLAG_ACCESS_IMPL_ADDR(t) \
+  &flag_access_ ## t,
+
+JVM_FLAG_NON_STRING_TYPES_DO(FLAG_ACCESS_IMPL_INIT)
+
+static const FlagAccessImpl* flag_accesss[JVMFlag::NUM_FLAG_TYPES] = {
+  JVM_FLAG_NON_STRING_TYPES_DO(FLAG_ACCESS_IMPL_ADDR)
+  // ccstr and ccstrlist have special setter
+};
+
+inline const FlagAccessImpl* JVMFlagAccess::access_impl(const JVMFlag* flag) {
+  int type = flag->type();
+  int max = (int)(sizeof(flag_accesss)/sizeof(flag_accesss[0]));
+  assert(type >= 0 && type < max , "sanity");
+
+  return flag_accesss[type];
+}
+
+// This is called by JVMFlagAccess::*AtPut() and JVMFlagAccess::set<...>(JVMFlag* flag, ...)
+JVMFlag::Error JVMFlagAccess::set_impl(JVMFlag* flag, int type_enum, void* value, JVMFlag::Flags origin) {
+  if (type_enum == JVMFlag::TYPE_ccstr || type_enum == JVMFlag::TYPE_ccstrlist) {
+    return ccstrAtPut(flag, (ccstr*)value, origin);
+  }
+
+  if (flag == NULL) {
+    return JVMFlag::INVALID_FLAG;
+  }
+  if (flag->type() != type_enum) {
+    return JVMFlag::WRONG_FORMAT;
+  }
+  return access_impl(flag)->set(flag, value, origin);
+}
+
+JVMFlag::Error JVMFlagAccess::ccstrAtPut(JVMFlag* flag, ccstr* value, JVMFlag::Flags origin) {
+  if (flag == NULL) return JVMFlag::INVALID_FLAG;
+  if (!flag->is_ccstr()) return JVMFlag::WRONG_FORMAT;
+  ccstr old_value = flag->get_ccstr();
+  trace_flag_changed<ccstr, EventStringFlagChanged>(flag, old_value, *value, origin);
+  char* new_value = NULL;
+  if (*value != NULL) {
+    new_value = os::strdup_check_oom(*value);
+  }
+  flag->set_ccstr(new_value);
+  if (flag->is_default() && old_value != NULL) {
+    // Prior value is NOT heap allocated, but was a literal constant.
+    old_value = os::strdup_check_oom(old_value);
+  }
+  *value = old_value;
+  flag->set_origin(origin);
+  return JVMFlag::SUCCESS;
+}
+
+// This is called by the SET_FLAG_XXX macros.
+JVMFlag::Error JVMFlagAccess::set_impl(JVMFlagsEnum flag_enum, int type_enum, void* value, JVMFlag::Flags origin) {
+  if (type_enum == JVMFlag::TYPE_ccstr || type_enum == JVMFlag::TYPE_ccstrlist) {
+    return ccstrAtPut((JVMFlagsEnum)flag_enum, *((ccstr*)value), origin);
+  }
+
+  JVMFlag* flag = JVMFlag::flag_from_enum(flag_enum);
+  assert(flag->type() == type_enum, "wrong flag type");
+  return set_impl(flag, type_enum, value, origin);
+}
+
+// This is called by the FLAG_SET_XXX macros.
```

Comment:
There are both FLAG_SET_XXX and SET_FLAG_XXX macros ?  Is this a typo in the comment?